src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java

```diff
@@ -144,32 +139,41 @@ public Repository getOrCheckoutRepository(URL repoUrl) throws IOException, Inter
                 // make sure that cloneInProgress is released
                 cloneInProgressOperations.remove(localPath);
             }
+            return getRepositoryByLocalPath(localPath);
         }
-        else {
-            log.debug("Repository at " + localPath + " already exists");
-            // in this case we pull for changes to make sure the Git repo is up to date
-            shouldPullChanges = true;
+    }
+
+    /**
+     * Get a git repository that is checked out on the server. Throws immediately an exception if the localPath does not exist. Will first try to retrieve a cached repository from
+     * cachedRepositories. Side effect: This method caches retrieved repositories in a HashMap, so continuous retrievals can be avoided (reduces load).
+     *
+     * @param localPath to git repo on server.
+     * @return the git repository in the localPath (if exists!)
+     * @throws IOException if the the filepath can't be found or does not contain a git repository.
+     */
+    private Repository getRepositoryByLocalPath(Path localPath) throws IOException {
+        if (!Files.exists(localPath)) {
+            // In this case we should remove the repository if cached, because it can't exist anymore.
+            cachedRepositories.remove(localPath);
+            throw new IOException();
```

Comment:
can you not just return null here?

Reply:
You are right, this is nicer. I also changed the JavaDocs of this method.